### PR TITLE
Shaders refactor

### PR DESCRIPTION
**Pull request by jfirebaugh**
Followup to #6468, specifically https://github.com/mapbox/mapbox-gl-native/pull/6468#discussion_r85246468.
- [x] Fix overdraw debug mode

### CODE REVIEW

**Comment by brunoabinader** [comment]
Potentially fixes the issues mentioned in https://github.com/mapbox/mapbox-gl-native/issues/6799#issue-184702365 (needs verification).

**Comment by jfirebaugh** [comment]
Ready for review!

**Comment by jfirebaugh** [comment]
@kkaefer The test sources now need the path for generated shader .hpp's in their include paths. Where in the cmake files do I change that?

**Comment by kkaefer** [comment]
You can add `${MBGL_GENERATED}/include` to the include dir list. The actual code that generates the shaders is in `cmake/shaders.cmake`.

**Comment by kkaefer** [comment]
I'd propose adding this to enforce initialization of the primitive types with a value:

```diff
diff --git a/src/mbgl/gl/draw_mode.hpp b/src/mbgl/gl/draw_mode.hpp
index 8f7db0a..b099a5f 100644
--- a/src/mbgl/gl/draw_mode.hpp
+++ b/src/mbgl/gl/draw_mode.hpp
@@ -11,6 +11,7 @@ public:
     using Primitive = Point;
     static constexpr std::size_t bufferGroupSize = 1;
 
+    Points(float pointSize_) : pointSize(pointSize_) {}
     float pointSize;
 };
 
@@ -19,6 +20,9 @@ public:
     using Primitive = Line;
     static constexpr std::size_t bufferGroupSize = 2;
 
+    Lines(float lineWidth_) : lineWidth(lineWidth_) {
+        assert(lineWidth > 0);
+    }
     float lineWidth;
 };
 
@@ -29,6 +33,9 @@ public:
     using Primitive = Line;
     static constexpr std::size_t bufferGroupSize = 1;
 
+    LineStrip(float lineWidth_) : lineWidth(lineWidth_) {
+        assert(lineWidth > 0);
+    }
     float lineWidth;
 };
 
```

**Comment by kkaefer** [comment]
One more thing that I'd like to see (but that is out of scope for this PR) is to somehow move shaders to the context. Background is that for Node.js, we'd like to use a shared context, and ideally we can share the shaders as well to avoid having all of these Programs that are just duplicates.

File: src/mbgl/util/indexed_tuple.hpp [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+        return std::get<Index<I>>(*this);
+    }
+
+    template <class I>
+    const auto& get() const {
+        return std::get<Index<I>>(*this);
+    }
+};
+
+} // namespace mgbl
```
Comment (jfirebaugh) [inline src/mbgl/util/indexed_tuple.hpp]: * Less verbose -- `std::get<typename As::SomeInnerType>(tuple)` becomes `tuple.template get<As>()`.
* Eliminates the need to use a separate helper and `std::index_sequence` in `Attributes::binder`.
* Will be consistent with use in #6868.

File: src/mbgl/gl/segment.hpp [hunk 2]
```diff
@@ -8,19 +8,19 @@ namespace gl {
...
-          primitiveLength(primitiveLength_) {}
+          indexLength(indexLength_) {}
 
     const std::size_t vertexOffset;
-    const std::size_t primitiveOffset;
+    const std::size_t indexOffset;
 
     std::size_t vertexLength;
-    std::size_t primitiveLength;
+    std::size_t indexLength;
```
Comment (jfirebaugh) [inline src/mbgl/gl/segment.hpp]: Fundamentally, a vertex/index buffer is a buffer of vertices/indexes, not a buffer of triangles, or lines, or whatever. Conceptually, that's how OpenGL works, and I found it easier to reason about things when I stopped trying to think in terms of buffers of primitives.

Implementation wise, it simplifies calculations like [this](https://github.com/mapbox/mapbox-gl-native/pull/6856/files#diff-617d0e0d315b4076d9fb82555ca6df0cL464).

It matches the internal implementation of the new `{Vertex,Index}Vector` -- I think if we were to have `std::vector<Triangle<Vertex>>`, the packing might come out wrong. You don't want to get in a situation where there might be extra padding every third vertex. `{Vertex,Index}Vector` still enforces filling the vector with the right sized groups though.

File: src/mbgl/programs/raster_program.hpp [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+                p.x,
+                p.y
+            },
+            {
+                t.x,
+                t.y
+            }
+        };
+    }
+};
```
Comment (kkaefer) [inline src/mbgl/programs/raster_program.hpp]: Can we factor out the attributes stuff into its own objects to be in line with the others?

**Comment by jfirebaugh** [comment]
Rebasing this is getting to be a real drag. Can I get this approved and we can follow up on nits in other PRs?

**Comment by brunoabinader** [comment]
Thank you @jfirebaugh - confirming that overdraw debug mode is operational again on Linux:
overdraw

**Pull request merged**